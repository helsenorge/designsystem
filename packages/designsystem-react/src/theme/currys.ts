import {PaletteNames, PaletteDepths} from './palette';
import {theme} from './';

const isNotPalette = (name: string) => name === 'white' || name === 'black';

export const getColor = (name: PaletteNames | 'white' | 'black', depth: PaletteDepths) =>
  isNotPalette(name) ? '#FFFFFF' : theme.palette[`${name}${depth}`];

export const getHoverColor = (color: PaletteNames | 'white' | 'black') => `${getColor(color, 700)}1A`;
