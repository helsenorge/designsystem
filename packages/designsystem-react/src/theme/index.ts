import {palette} from './palette';
import {typography} from './typography';
import {spacer, spacers} from './spacers';
import {screen, breakpoints} from './grid';

export const theme = {
  spacer,
  spacers,
  palette,
  typography,
  screen,
  breakpoints,
};
