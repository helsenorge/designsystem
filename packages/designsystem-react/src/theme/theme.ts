import {palette} from './palette';

export const theme = {
  palette,
  animation: {
    duration: '205ms',
  },
};
