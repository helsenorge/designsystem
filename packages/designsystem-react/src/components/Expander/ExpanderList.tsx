import React, {useState, useEffect} from 'react';
import {
  StyledExpanderList,
  StyledExpanderListLink,
  StyledExpanderListLinkContent,
  StyledExpanderContent,
} from './ExpanderList.styled';
import {PaletteNames} from '../../theme/palette';
import Icon from '../Icons';

export type ExpanderListColors = PaletteNames;

export interface CompoundComponent
  extends React.ForwardRefExoticComponent<ExpanderListProps & React.RefAttributes<HTMLUListElement>> {
  Expander: any;
}

interface ExpanderListProps {
  accordion?: boolean;
  bottomBorder?: boolean;
  children: React.ReactNode;
  color: ExpanderListColors;
  large?: boolean;
  topBorder?: boolean;
}

interface ExpanderProps extends React.HTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode;
  color: ExpanderListColors;
  icon?: React.ReactElement;
  isExpanded?: boolean;
  large: boolean;
  title: string;
}

const Expander = React.forwardRef((props: ExpanderProps, ref: any) => {
  const {children, color, icon, large, title, isExpanded = false, ...restProps} = props;
  return (
    <li>
      <StyledExpanderListLink
        isExpanded={isExpanded}
        hasIcon={!!icon}
        large={large}
        color={color}
        ref={ref}
        {...restProps}>
        <StyledExpanderListLinkContent>
          {icon && React.cloneElement(icon, {size: 48})}
          {title}
        </StyledExpanderListLinkContent>
        <Icon type={isExpanded ? 'chevronUp' : 'chevronDown'} />
      </StyledExpanderListLink>
      {isExpanded ? <StyledExpanderContent>{children}</StyledExpanderContent> : null}
    </li>
  );
});

const ExpanderList = React.forwardRef((props: ExpanderListProps, ref: any) => {
  const {children, large, color, accordion = false, topBorder = true, bottomBorder = true} = props;
  const [activeExpander, setActiveExpander] = useState({});
  function handleExpanderClick(e: any) {
    const id = e.currentTarget.id;
    setActiveExpander(prevState =>
      accordion ? {[id]: !Boolean(prevState[id])} : {...prevState, [id]: !Boolean(prevState[id])},
    );
  }
  return (
    <StyledExpanderList topBorder={topBorder} bottomBorder={bottomBorder} ref={ref}>
      {React.Children.map(children, (child: any, index: number) => {
        if (child.type.displayName === 'ExpanderList.Expander') {
          const isExpanded = activeExpander[index];
          return React.cloneElement(child, {
            id: index,
            key: index,
            isExpanded,
            color,
            large,
            onClick: handleExpanderClick,
          });
        }
      })}
    </StyledExpanderList>
  );
}) as CompoundComponent;

ExpanderList.Expander = Expander;
Expander.displayName = 'ExpanderList.Expander';

export default ExpanderList;
