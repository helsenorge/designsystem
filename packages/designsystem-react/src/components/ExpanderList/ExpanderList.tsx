import React, {useState} from 'react';
import {
  StyledExpanderList,
  StyledExpanderListLink,
  StyledExpanderListLinkContent,
  StyledExpanderContent,
  StyledExpanderListIconContainer,
} from './ExpanderList.styled';
import {PaletteNames} from '../../theme/palette';
import Icon from '../Icons';
import ChevronUp from '../Icons/ChevronUp';
import ChevronDown from '../Icons/ChevronDown';
import {useHover} from '../../hooks/useHover';

export type ExpanderListColors = PaletteNames;
export type ExpanderType = React.ForwardRefExoticComponent<ExpanderProps & React.RefAttributes<HTMLLIElement>>;

export interface ExpanderListCompound
  extends React.ForwardRefExoticComponent<ExpanderListProps & React.RefAttributes<HTMLUListElement>> {
  Expander: ExpanderType;
}

interface MouseEventWithPath extends MouseEvent {
  path?: Array<HTMLElement>;
}

interface ExpanderListProps {
  accordion?: boolean;
  bottomBorder?: boolean;
  children: React.ReactNode;
  childPadding?: boolean;
  className?: string;
  color?: ExpanderListColors;
  large?: boolean;
  isOpen?: boolean;
  topBorder?: boolean;
}

type Modify<T, R> = Omit<T, keyof R> & R;

type ExpanderProps = Modify<
  React.HTMLAttributes<HTMLButtonElement>,
  {
    title: JSX.Element | string;
    children: React.ReactNode;
    className?: string;
    color?: ExpanderListColors;
    icon?: React.ReactElement;
    padding?: boolean;
    isExpanded?: boolean;
    large?: boolean;
    testId?: string;
  }
>;

// TODO: See what can be done with regards to double reffing.
const Expander: ExpanderType = React.forwardRef((props: ExpanderProps, ref: React.Ref<HTMLLIElement>) => {
  const {
    children,
    padding = true,
    color = 'neutral',
    className = '',
    icon,
    large = false,
    title,
    isExpanded = false,
    testId = '',
    ...restProps
  } = props;
  const {hoverRef, isHovered} = useHover<HTMLButtonElement>();
  return (
    <li ref={ref}>
      <StyledExpanderListLink
        data-testid={testId}
        className={className}
        isExpanded={isExpanded}
        large={large}
        color={color}
        ref={hoverRef}
        {...restProps}>
        <StyledExpanderListLinkContent>
          {icon ? (
            <StyledExpanderListIconContainer>
              {React.cloneElement(icon, {size: 48, isHovered})}
            </StyledExpanderListIconContainer>
          ) : null}
          {title}
        </StyledExpanderListLinkContent>
        <StyledExpanderListIconContainer>
          <Icon svgIcon={isExpanded ? ChevronUp : ChevronDown} isHovered={isHovered} />
        </StyledExpanderListIconContainer>
      </StyledExpanderListLink>
      {isExpanded ? <StyledExpanderContent padding={padding}>{children}</StyledExpanderContent> : null}
    </li>
  );
});

function findShadowDOMId(event: MouseEventWithPath, tagName: string): string {
  const pathElements: Array<HTMLElement> =
    event.path || (event.composedPath && (event.composedPath() as Array<HTMLElement>));
  const element = pathElements.find((el: HTMLElement) => el.tagName === tagName);
  return element ? element.id : '';
}

const ExpanderList = React.forwardRef((props: ExpanderListProps, ref: React.Ref<HTMLUListElement>) => {
  const {
    children,
    childPadding = true,
    large,
    isOpen = false,
    color,
    className = '',
    accordion = false,
    topBorder = true,
    bottomBorder = true,
  } = props;
  const [activeExpander, setActiveExpander] = useState({});

  function handleExpanderClick(event: React.MouseEvent<HTMLElement, MouseEvent>): void {
    const id = event.currentTarget?.id || findShadowDOMId((event as unknown) as MouseEventWithPath, 'BUTTON');
    if (!isOpen) {
      setActiveExpander(prevState => (accordion ? {[id]: !prevState[id]} : {...prevState, [id]: !prevState[id]}));
    }
  }

  return (
    <StyledExpanderList className={className} topBorder={topBorder} bottomBorder={bottomBorder} ref={ref}>
      {React.Children.map(children, (child: React.ReactNode, index: number) => {
        if ((child as React.ReactElement<ExpanderProps>).type === Expander) {
          const isExpanded = isOpen || activeExpander[`expander-${index}`];
          return React.cloneElement(child as React.ReactElement<ExpanderProps>, {
            id: `expander-${index}`,
            key: `expander-${index}`,
            isExpanded,
            padding: childPadding,
            color,
            large,
            'aria-expanded': isExpanded,
            onClick: handleExpanderClick,
          });
        }
        return child;
      })}
    </StyledExpanderList>
  );
}) as ExpanderListCompound;

ExpanderList.Expander = Expander;
Expander.displayName = 'ExpanderList.Expander';

export default ExpanderList;
