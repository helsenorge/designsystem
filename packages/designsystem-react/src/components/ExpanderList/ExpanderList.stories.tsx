import React from 'react';
import {withKnobs, text, select, boolean} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import ExpanderList from './ExpanderList';
import LinkList from '../LinkList';
import Title from '../Title';
import Icon from '../Icons';
import Avatar from '../Icons/Avatar';
import PaperPlane from '../Icons/PaperPlane';
import AlarmClock from '../Icons/AlarmClock';
import {allPaletteNames} from '../../../.storybook/knobs';

const stories = storiesOf('ExpanderList', module);
stories.addDecorator(withKnobs);

stories.add('Default', () => (
  <div
    style={{
      width: '40rem',
    }}>
    <ExpanderList
      isOpen={boolean('Is open', false)}
      accordion={boolean('Accordion', false)}
      childPadding={boolean('Child padding', true)}
      color={select('Color', allPaletteNames, 'blueberry')}>
      <ExpanderList.Expander title="Cognitive therapy">
        Some major content/text about <a href="#">certain illness here.</a>. Note that the anchors are not styled by
        default. The content is free from styling.
      </ExpanderList.Expander>
      <ExpanderList.Expander title="Health anxiety">
        Some major content/text about certain illness here.
      </ExpanderList.Expander>
      <ExpanderList.Expander title="Brain damage (This is a super long text just so you can observe how it will wrap on a multiline context)">
        Some major content/text about certain illness here.
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));

stories.add('With icon', () => (
  <div
    style={{
      width: '40rem',
    }}>
    <ExpanderList isOpen={boolean('Is open', false)} color="cherry">
      <ExpanderList.Expander icon={<Icon svgIcon={Avatar} />} title="Cognitive therapy">
        Some major content/text about certain illness here.
      </ExpanderList.Expander>
      <ExpanderList.Expander icon={<Icon svgIcon={PaperPlane} />} title="Health anxiety">
        Some major content/text about certain illness here.
      </ExpanderList.Expander>
      <ExpanderList.Expander icon={<Icon svgIcon={AlarmClock} />} title="Brain damage">
        Some major content/text about certain illness here.
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));

stories.add('With LinkList', () => (
  <div
    style={{
      width: '40rem',
    }}>
    <ExpanderList isOpen={boolean('Is open', false)} color="cherry">
      <ExpanderList.Expander title="Cognitive therapy">
        <LinkList color="cherry">
          <LinkList.Link href="/firstaid">First aid</LinkList.Link>
          <LinkList.Link href="/secondaid">Second aid</LinkList.Link>
          <LinkList.Link href="/thirdaid">Third aid</LinkList.Link>
        </LinkList>
      </ExpanderList.Expander>
    </ExpanderList>
  </div>
));
