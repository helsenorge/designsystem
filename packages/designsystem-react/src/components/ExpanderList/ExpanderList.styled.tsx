import styled, {css} from 'styled-components';
import {getColor} from '../../theme/currys';
import {ExpanderListColors} from './ExpanderList';

interface StyledExpanderListProps {
  topBorder?: boolean;
  bottomBorder?: boolean;
}

interface StyledExpanderContentProps {
  padding: boolean;
}

const topBorder = (props: StyledExpanderListProps) =>
  props.topBorder &&
  css`
    & > :first-child {
      border-top: 1px solid ${getColor('neutral', 500)};
    }
  `;

const bottomBorder = (props: StyledExpanderListProps) =>
  !props.bottomBorder &&
  css`
    & > :last-child {
      border-bottom: 0;
    }
  `;

const largeStyle = (props: StyledExpanderListLinkProps) =>
  props.large &&
  css`
    font-size: 1.5rem;
    line-height: 2rem;
  `;

const StyledExpanderList = styled('ul')`
  list-style: none;
  padding: 0;
  width: inherit;
  margin: 0;
  & > :nth-child(n) {
    border-bottom: 1px solid ${getColor('neutral', 500)};
  }
  ${topBorder}
  ${bottomBorder}
`;

const StyledExpanderContent = styled('div')<StyledExpanderContentProps>`
  padding: ${props => (props.padding ? '1.5rem 0.5rem' : '0')};
`;

const StyledExpanderListIconContainer = styled('span')`
  width: 3rem;
  .hnds-style-icon {
    vertical-align: middle;
  }
`;

// TODO: Need to overhaul all icons to be Styled insteadof hnds-style-icon.
const StyledExpanderListLinkContent = styled('span')`
  display: flex;
  align-items: center;
  text-align: left;
  border-bottom: 1px solid transparent;
  ${StyledExpanderListIconContainer} {
    margin-right: 0.5rem;
  }
`;

interface StyledExpanderListLinkProps {
  color: ExpanderListColors;
  large: boolean;
  isExpanded: boolean;
}

const StyledExpanderListLink = styled('button')<StyledExpanderListLinkProps>`
  padding: 0.5rem;
  display: flex;
  cursor: pointer;
  font-size: 1.25rem;
  line-height: 1.875rem;
  text-decoration: none;
  justify-content: space-between;
  align-items: center;
  letter-spacing: unset;
  color: ${getColor('black')};
  background-color: ${props =>
    props.isExpanded
      ? props.color === 'white'
        ? getColor('neutral', 50)
        : getColor(props.color, 100)
      : 'transparent'};
  border: 0;
  width: 100%;

  outline: none;
  ${largeStyle};
  &:hover,
  :focus {
    background-color: ${props => (props.color === 'white' ? getColor('neutral', 50) : getColor(props.color, 100))};
  }
  &:focus {
    outline: none;
    ${StyledExpanderListLinkContent} {
      border-color: ${getColor('black')};
    }
  }
`;

export {
  StyledExpanderList,
  StyledExpanderListLink,
  StyledExpanderListLinkContent,
  StyledExpanderContent,
  StyledExpanderListIconContainer,
};
