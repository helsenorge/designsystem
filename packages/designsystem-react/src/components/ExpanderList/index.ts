import ExpanderList from './ExpanderList';
export default ExpanderList;
