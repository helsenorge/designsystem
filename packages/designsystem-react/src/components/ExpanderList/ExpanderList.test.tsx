import React from 'react';
import {render, fireEvent, screen, act} from '@testing-library/react';
import ExpanderList from './ExpanderList';
import {theme} from '../../theme';

describe('Gitt ExpanderList blir rendret', (): void => {
  test('Sjekk at ExpanderList sitt snapshot matcher', (): void => {
    const {container} = render(
      <div
        style={{
          width: '40rem',
        }}>
        <ExpanderList isOpen={false} accordion={false} childPadding={true} color={'blueberry'}>
          <ExpanderList.Expander title="Title 1">Text 1</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">Text 2</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">Text 2</ExpanderList.Expander>
        </ExpanderList>
      </div>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Sjekk Expander bakgrunnsfarge når musepeker trykker på den', (): void => {
    render(
      <div
        style={{
          width: '40rem',
        }}>
        <ExpanderList isOpen={false} accordion={false} childPadding={true} color={'blueberry'}>
          <ExpanderList.Expander testId="test01" title="Title 1">
            Text 1
          </ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">Text 2</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">Text 2</ExpanderList.Expander>
        </ExpanderList>
      </div>,
    );

    const testExpander = screen.getByTestId('test01');
    fireEvent.click(testExpander);
    expect(testExpander).toHaveStyle(`background-color: ${theme.palette.blueberry100}`);
  });
});
