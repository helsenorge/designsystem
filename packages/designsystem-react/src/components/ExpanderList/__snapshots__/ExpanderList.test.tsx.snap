// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt ExpanderList blir rendret Sjekk at ExpanderList sitt snapshot matcher 1`] = `
<div>
  <div
    style="width: 40rem;"
  >
    <ul
      class="sc-AxjAm cdEKGT"
    >
      <li>
        <button
          class="sc-AxhUy eNACsH"
          color="blueberry"
          data-testid=""
          id="expander-0"
        >
          <span
            class="sc-AxhCb gPRmPx"
          >
            Title 1
          </span>
          <span
            class="sc-AxiKw hcQbGl"
          >
            <svg
              aria-hidden="true"
              class="sc-AxgMl gUSQvJ hnds-style-icon "
              fill="black"
              height="48"
              role="img"
              viewBox="0 0 48 48"
              width="48"
            >
              <path
                d="M32.951 18.788l1.419 1.409-10.371 10.451L13.63 20.197l1.419-1.409L24 27.809z"
              />
            </svg>
          </span>
        </button>
      </li>
      <li>
        <button
          class="sc-AxhUy eNACsH"
          color="blueberry"
          data-testid=""
          id="expander-1"
        >
          <span
            class="sc-AxhCb gPRmPx"
          >
            Title 2
          </span>
          <span
            class="sc-AxiKw hcQbGl"
          >
            <svg
              aria-hidden="true"
              class="sc-AxgMl gUSQvJ hnds-style-icon "
              fill="black"
              height="48"
              role="img"
              viewBox="0 0 48 48"
              width="48"
            >
              <path
                d="M32.951 18.788l1.419 1.409-10.371 10.451L13.63 20.197l1.419-1.409L24 27.809z"
              />
            </svg>
          </span>
        </button>
      </li>
      <li>
        <button
          class="sc-AxhUy eNACsH"
          color="blueberry"
          data-testid=""
          id="expander-2"
        >
          <span
            class="sc-AxhCb gPRmPx"
          >
            Title 3
          </span>
          <span
            class="sc-AxiKw hcQbGl"
          >
            <svg
              aria-hidden="true"
              class="sc-AxgMl gUSQvJ hnds-style-icon "
              fill="black"
              height="48"
              role="img"
              viewBox="0 0 48 48"
              width="48"
            >
              <path
                d="M32.951 18.788l1.419 1.409-10.371 10.451L13.63 20.197l1.419-1.409L24 27.809z"
              />
            </svg>
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
